samza-core/src/main/java/org/apache/samza/operators/MessageStreamImpl.java

```diff
@@ -65,52 +67,59 @@
    *
    * @param graph the {@link StreamGraphImpl} object that this stream belongs to
    */
-  MessageStreamImpl(StreamGraphImpl graph) {
+  public MessageStreamImpl(StreamGraphImpl graph) {
     this.graph = graph;
   }
 
   @Override
   public <TM> MessageStream<TM> map(MapFunction<M, TM> mapFn) {
-    OperatorSpec<TM> op = OperatorSpecs.<M, TM>createMapOperatorSpec(mapFn, this.graph, new MessageStreamImpl<>(this.graph));
+    OperatorSpec<TM> op = OperatorSpecs.<M, TM>createMapOperatorSpec(
+        mapFn, new MessageStreamImpl<>(this.graph), this.graph.getNextOpId());
     this.registeredOperatorSpecs.add(op);
     return op.getNextStream();
   }
 
   @Override
   public MessageStream<M> filter(FilterFunction<M> filterFn) {
-    OperatorSpec<M> op = OperatorSpecs.<M>createFilterOperatorSpec(filterFn, this.graph, new MessageStreamImpl<>(this.graph));
+    OperatorSpec<M> op = OperatorSpecs.<M>createFilterOperatorSpec(
+        filterFn, new MessageStreamImpl<>(this.graph), this.graph.getNextOpId());
     this.registeredOperatorSpecs.add(op);
     return op.getNextStream();
   }
 
   @Override
   public <TM> MessageStream<TM> flatMap(FlatMapFunction<M, TM> flatMapFn) {
-    OperatorSpec<TM> op = OperatorSpecs.createStreamOperatorSpec(flatMapFn, this.graph, new MessageStreamImpl<>(this.graph));
+    OperatorSpec<TM> op = OperatorSpecs.createStreamOperatorSpec(
+        flatMapFn, new MessageStreamImpl<>(this.graph), this.graph.getNextOpId());
     this.registeredOperatorSpecs.add(op);
     return op.getNextStream();
```

Comment:
There's a (too?) subtle distinction b/w `nextStream` and `outputStream` in this PR - `nextStream` is used for the next MessageStream to propagate results to, while `outputStream` means an actual/physical output stream. If the usage is not consistent please LMK. For example, the `outputStream` to sendTo et. al is _not_ the nextStream for the SinkOperator since it's terminal.